core/include/scipp/core/dtype.h

```diff
@@ -32,71 +23,13 @@ template <class T>
 struct is_sparse<const sparse_container<T> &> : std::true_type {};
 template <class T> inline constexpr bool is_sparse_v = is_sparse<T>::value;
 
-enum class DType {
-  Double,
-  Float,
-  Int32,
-  Int64,
-  String,
-  Bool,
-  SparseDouble,
-  SparseFloat,
-  SparseInt64,
-  SparseInt32,
-  SparseBool,
-  DataArray,
-  Dataset,
-  EigenVector3d,
-  EigenQuaterniond,
-  PyObject,
-  SpanDouble,
-  SpanFloat,
-  SpanInt64,
-  SpanInt32,
-  SpanConstDouble,
-  SpanConstFloat,
-  SpanConstInt64,
-  SpanConstInt32,
-  Unknown
+struct SCIPP_CORE_EXPORT DType {
+  std::type_index index;
```

Comment:
should be const i think. Can't see any reason we would want this to change.

Reply:
We need `DType` to be assignable and copyable. Relevant classes like `Variable` never expose a mutable reference to their `DType`, so changing it does not harm.